File: api/api/webhook.js
const TELEGRAM_TOKEN   = TELEGRAM_TOKEN_ENV   // variables de entorno en CF
const TELEGRAM_CHAT_ID = TELEGRAM_CHAT_ID_ENV
const GH_PAT           = GH_PAT_ENV
const GITHUB_REPO      = "DevCop95/dev101_bot"

async function sendMessage(chatId, text) {
  await fetch(`https://api.telegram.org/bot${TELEGRAM_TOKEN}/sendMessage`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ chat_id: chatId, text, parse_mode: "Markdown" })
  })
}

async function triggerGithubAction() {
  const res = await fetch(
    `https://api.github.com/repos/${GITHUB_REPO}/actions/workflows/bot.yml/dispatches`,
    {
      method: "POST",
      headers: {
        "Authorization": `token ${GH_PAT}`,
        "Accept": "application/vnd.github.v3+json",
        "Content-Type": "application/json"
      },
      body: JSON.stringify({ ref: "main" })
    }
  )
  return res.status === 204
}

export default {
  async fetch(request) {
    if (request.method !== "POST") {
      return new Response("OK", { status: 200 })
    }

    const body     = await request.json()
    const message  = body?.message || {}
    const chatId   = String(message?.chat?.id || "")
    const text     = (message?.text || "").trim()

    // ── Bloquea cualquier usuario no autorizado ───────────────────────────
    if (chatId !== TELEGRAM_CHAT_ID) {
      await sendMessage(chatId, "⛔ No estás autorizado para usar este bot.")
      return new Response("OK", { status: 200 })
    }

    // ── Comandos ──────────────────────────────────────────────────────────
    if (text === "/noticias") {
      const ok = await triggerGithubAction()
      await sendMessage(chatId, ok
        ? "🚀 Buscando noticias... llegan en unos segundos."
        : "❌ Error al lanzar el job. Revisa GitHub Actions."
      )

    } else if (text === "/start" || text === "/help") {
      await sendMessage(chatId,
        "👾 *dev101\\_bot*\n\n" +
        "Comandos:\n" +
        "/noticias — busca y envía noticias ahora\n" +
        "/help — muestra este mensaje\n\n" +
        "_Envío automático cada 3 horas vía GitHub Actions._"
      )

    } else {
      await sendMessage(chatId, "Comando no reconocido. Usa /help.")
    }

    return new Response("OK", { status: 200 })
  }
}

File: api/webhook.py
from http.server import BaseHTTPRequestHandler
import json, os, requests

TELEGRAM_TOKEN   = os.environ["TELEGRAM_TOKEN"]
TELEGRAM_CHAT_ID = os.environ["TELEGRAM_CHAT_ID"]   # solo tú puedes usar el bot
GITHUB_TOKEN     = os.environ["GH_PAT"]
GITHUB_REPO      = "DevCop95/dev101_bot"

def trigger_github_action():
    """Dispara el workflow de noticias manualmente vía GitHub API."""
    url = f"https://api.github.com/repos/{GITHUB_REPO}/actions/workflows/bot.yml/dispatches"
    r = requests.post(url, headers={
        "Authorization": f"token {GITHUB_TOKEN}",
        "Accept": "application/vnd.github.v3+json"
    }, json={"ref": "main"})
    return r.status_code == 204

def send_message(chat_id, text):
    requests.post(
        f"https://api.telegram.org/bot{TELEGRAM_TOKEN}/sendMessage",
        json={"chat_id": chat_id, "text": text}
    )

class handler(BaseHTTPRequestHandler):

    def do_POST(self):
        length  = int(self.headers.get("Content-Length", 0))
        body    = json.loads(self.rfile.read(length))

        message = body.get("message", {})
        chat_id = str(message.get("chat", {}).get("id", ""))
        text    = message.get("text", "").strip()

        # ── Solo responde al dueño ────────────────────────────────────────
        if chat_id != TELEGRAM_CHAT_ID:
            send_message(chat_id, "⛔ No estás autorizado para usar este bot.")
            self._ok()
            return

        # ── Comandos ──────────────────────────────────────────────────────
        if text == "/noticias":
            ok = trigger_github_action()
            if ok:
                send_message(chat_id, "🚀 Buscando noticias... llegará en unos segundos.")
            else:
                send_message(chat_id, "❌ Error al lanzar el job. Revisa los logs en GitHub Actions.")

        elif text == "/start" or text == "/help":
            send_message(chat_id, (
                "👾 *dev101_bot*\n\n"
                "Comandos disponibles:\n"
                "/noticias — busca y envía noticias ahora\n"
                "/help — muestra este mensaje\n\n"
                "_El bot también envía noticias automáticamente cada 3 horas._"
            ))

        else:
            send_message(chat_id, "Comando no reconocido. Usa /help.")

        self._ok()

    def _ok(self):
        self.send_response(200)
        self.end_headers()
        self.wfile.write(b"OK")

    def log_message(self, *args):
        pass  # silencia logs de acceso
